File: activity.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

using std::cin;
using std::cout;
using std::endl;
using std::ifstream;
using std::istringstream;
using std::ofstream;
using std::string;
using std::vector;

struct activity {
  int activityNumber;
  int startTime;
  int finishTime;
  bool inOptimal = false;
};

vector<activity> schedule(vector<activity>* activities);
void printSchedule(vector<activity>* schedule, int* setNumber);
vector<activity> mergeSort(vector<activity>* list);
vector<activity> merge(vector<activity>* list1, vector<activity>* list2);

int main() {

  ifstream inputFile;
  ofstream outputFile;
  string inFileName, outFileName, tempString;
  istringstream stream;
  vector<activity> activities, optimalSchedule;
  int set = 1, toSort;

  // Prompt for a file name until the specified name can be opened
  do {
    std::cout << "Please enter the name and extension of a text file to read: " << endl;
    getline(cin, inFileName);
    inputFile.open(inFileName);
  } while (!inputFile);

  while (getline(inputFile, tempString)) {
    toSort = stoi(tempString);

    while (toSort > 0) {
      getline(inputFile, tempString);
      stream.str(tempString);

      activities.push_back(activity());
      stream >> activities.at(activities.size() - 1).activityNumber;
      stream >> activities.at(activities.size() - 1).startTime;
      stream >> activities.at(activities.size() - 1).finishTime;

      stream.clear();
      toSort--;
    }

    // Sort activities by start time in descending order
    activities = mergeSort(&activities);

    // Pass the sorted vector by reference to get an optimal schedule
    optimalSchedule = schedule(&activities);

    // Output an optimal schedule to the screen
    printSchedule(&optimalSchedule, &set);

    // Empty the vectors to prepare for the next set of activities to schedule
    activities.clear();
    optimalSchedule.clear();
    set++;

  }

  inputFile.close();
  return 0;

}

vector<activity> schedule(vector<activity>* activities) {
  vector<activity> optimalSchedule;

  // Start by adding the activity with the latest start time to our schedule
  optimalSchedule.push_back(activities->at(0));
  activities->at(0).inOptimal = true;

  // Track the most recently added activity in variable i, loop through the
  // remaining activities, and continue adding the activity with the next latest
  // start time that doesn't have an overlapping finish time with our most
  // recently added activity. Update i to the index of a newly added activity
  // so we're comparing with the most recently added activity for each new
  // addition to the schedule
  int i = 0;
  for (int j = 1; j < activities->size(); j++) {
    if (activities->at(j).finishTime <= activities->at(i).startTime) {
      optimalSchedule.push_back(activities->at(j));
      activities->at(j).inOptimal = true;
      i = j;
    }
  }

  return optimalSchedule;

}

void printSchedule(vector<activity>* schedule, int* setNumber) {
  cout << "Set " << *setNumber << endl;
  cout << "Number of activities selected: " << schedule->size() << endl;
  cout << "Activities: ";

  for (int i = schedule->size() - 1; i >= 0; i--) {
    cout << schedule->at(i).activityNumber;
    if (i != 0) {
      cout << ",";
    }
    cout << " ";
  }

  cout << endl << endl;
}

vector<activity> mergeSort(vector<activity>* list) {
  if (list->size() == 1) {
    // Handle the base case
    return *list;
  } else {
    // Split the vector into two vectors, then call mergeSort on each half
    int half = list->size() / 2;

    vector<activity> firstHalf (list->begin(), list->begin() + half);
    vector<activity> secondHalf (list->begin() + half, list->end());
    firstHalf = mergeSort(&firstHalf);
    secondHalf = mergeSort(&secondHalf);
    return merge(&firstHalf, &secondHalf);
  }
}

vector<activity> merge(vector<activity>* list1, vector<activity>* list2) {
  // Return the entirety of one of our vectors if the other is empty
  if (list1->size() == 0) {
    return *list2;
  }
  if (list2->size() == 0) {
    return *list1;
  }

  vector<activity> temp;
  int i = 0, j = 0;

  while (i < list1->size() && j < list2->size()) {
    if (list1->at(i).startTime > list2->at(j).startTime) {
      temp.push_back(list1->at(i));
      i++;
    } else {
      temp.push_back(list2->at(j));
      j++;
    }
  }

  // After exhausting one vector, add any remaining numbers to the end of our
  // temporary vector
  while (i < list1->size()) {
    temp.push_back(list1->at(i));
    i++;
  }
  while (j < list2->size()) {
    temp.push_back(list2->at(j));
    j++;
  }
  // Return our vector after iterating through both of the passed vectors
  return temp;
}

File: activity.py
#!/usr/bin/env python3

from mergesort import mergeSort


def openFile(fileName):
    with open(fileName, 'r') as file:
        contents = file.readlines()
    return contents


def schedule(activities):
    # Start by adding the event with the latest start time to our optimal schedule
    optimalSchedule = [activities[0]]
    # Initialize the most recently selected activity for our schedule, so we can track its starting
    # time and compare that to the end time of other elements sorted by their latest start time
    lastSelected = activities[0]
    for i, activity in enumerate(activities):
        # If an activity doesn't overlap with our most recent addition to the schedule, add it to
        # the schedule and set it as our most recently added activity to the basis of our comparison
        # for a non-overlapping activity
        if activity['endTime'] <= lastSelected['startTime']:
            optimalSchedule.append(activities[i])
            lastSelected = activity

    return optimalSchedule


def printSchedule(activities, setNumber):
    print('Set %d' % setNumber)
    print('Number of activities selected = %d' % len(activities))
    print('Activities: ', end='')
    for activity in activities:
        print(activity['activityNumber'], end=' ')
    print('\n')


def main():
    activityFile = openFile('act.txt')
    readLine = 0
    setNumber = 1

    while readLine < len(activityFile):
        # Get the number of lines in our text file, whether a line is listing the number of
        # activities in a set or describing a specific activity
        activitySetSize = int(activityFile[readLine])
        # Increment the number of lines we've read from the activity text file
        readLine += 1
        # Initialize an empty list of activities
        activitySet = []

        for activity in range(activitySetSize):
            # Initialize an empty dictionary so we can map the integers on each line to specific keys
            activityDict = {}
            # Convert each line into a list each integer as its own element
            activity = activityFile[readLine].split()
            # Convert the data type to integer for each element in the list (readlines() in openFile
            # gave us strings to start)
            activity = list(map(int, activity))
            # Make sure each value from each line corresponds to the correct info about the activity
            activityDict['activityNumber'] = activity[0]
            activityDict['startTime'] = activity[1]
            activityDict['endTime'] = activity[2]
            # Add the activity to our complete list of activities to schedule
            activitySet.append(activityDict)
            # Increment the number of lines we've read after we've completed parsing the data
            readLine += 1

        # Use merge sort to return a list of dictionaries based on the the start time key in
        # descending order
        activitySet = mergeSort(activitySet, 'startTime', False)
        # Pass the activity list to our optimizeSchedule function to get a list of activities
        # that don't overlap based on their latest start times
        optimalSchedule = schedule(activitySet)
        # Output the formatted results of the optimized schedule to the terminal
        printSchedule(optimalSchedule, setNumber)
        # Increment the number of activity sets we've processed during the while loop
        setNumber += 1


if __name__ == "__main__":
    main()
